core/modules/replica/PurgeJob.cc

```diff
@@ -0,0 +1,520 @@
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "restart");
+
+    if (_findAllJob or (_numLaunched != _numFinished)) {
+        throw std::logic_error ("PurgeJob::restart ()  not allowed in this object state");
+    }
+    _jobs.clear();
+
+    _numFailedLocks = 0;
+
+    _numLaunched = 0;
+    _numFinished = 0;
+}
+
+void PurgeJob::notifyImpl() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "notifyImpl");
+
+    if (_onFinish) {
+        _onFinish(shared_from_base<PurgeJob>());
+    }
+}
+
+void PurgeJob::onPrecursorJobFinish() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "onPrecursorJobFinish");
+
+    // IMPORTANT: the final state is required to be tested twice. The first time
+    // it's done in order to avoid deadlock on the "in-flight" requests reporting
+    // their completion while the job termination is in a progress. And the second
+    // test is made after acquering the lock to recheck the state in case if it
+    // has transitioned while acquering the lock.
+
+    if (state() == State::FINISHED) return;
+
+    util::Lock lock(_mtx, context() + "onPrecursorJobFinish");
+
+    if (state() == State::FINISHED) return;
+
+    ////////////////////////////////////////////////////////////////
+    // Only proceed with the replication effort if the precursor job
+    // has succeeded.
+
+    if (_findAllJob->extendedState() != ExtendedState::SUCCESS) {
+        finish(lock, ExtendedState::FAILED);
+        return;
+    }
+
+    /////////////////////////////////////////////////////////////////
+    // Analyse results and prepare a deletion plan to remove extra
+    // replocas for over-represented chunks
+    //
+    // IMPORTANT:
+    //
+    // - chunks which were found locked by some other job will not be deleted
+    //
+    // - when deciding on a number of replicas to be deleted the algorithm
+    //   will only consider 'good' chunks (the ones which meet the 'colocation'
+    //   requirement and which has good chunks only.
+    //
+    // - at a presense of more than one candidate for deletion, a worker with
+    //   more chunks will be chosen.
+    //
+    // - the statistics for the number of chunks on each worker will be
+    //   updated as deletion jobs targeting the corresponding
+    //   workers were issued.
+    //
+    // ATTENTION: the read-only workers will not be considered by
+    //            the algorithm. Those workers are used by different kinds
+    //            of jobs.
+
+    FindAllJobResult const& replicaData = _findAllJob->getReplicaData();
+
+    // The number of replicas to be deleted for eligible chunks
+    //
+    std::map<unsigned int,int> chunk2numReplicas2delete;
+
+    for (auto&& chunk2workers: replicaData.isGood) {
+        unsigned int const  chunk    = chunk2workers.first;
+        auto         const& replicas = chunk2workers.second;
+
+        // skip the special chunk which must be present on all workers
+        if (chunk == 1234567890) continue;
```

Comment:
`1234567890` should be defined as a constant somewhere and that named variable should be used here.

Reply:
I 100% agree with you. This will get improved with class ChunkNumber (which also supports chunk number validation based on a particular partitioning scheme). Please, have a look at:
ChunkNumber.h
ChunkNumber.cc
testChunkNumber.cc

This (new) design will also solve a problem of the "overflow" chunks. The actual number will be hidden behind the class interface.

I decided to postpone code migration to ChunkNumber in a separate ticket.
Let's discuss tin in person tomorrow.